core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -295,17 +297,44 @@ public static void writeShortDecimal(BlockBuilder blockBuilder, long value)
     public static long rescale(long value, int fromScale, int toScale)
     {
         if (toScale < fromScale) {
-            throw new IllegalArgumentException("target scale must be larger than source scale");
+            // Use MAX_SHORT_PRECISION to fit all precision long values
+            return shortToShortCast(
+                    value,
+                    MAX_SHORT_PRECISION,
+                    fromScale,
+                    MAX_SHORT_PRECISION,
+                    toScale,
+                    longTenToNth(fromScale - toScale),
+                    longTenToNth(fromScale - toScale - 1) * 5);
         }
-        return value * longTenToNth(toScale - fromScale);
+
+        long returnValue = value * longTenToNth(toScale - fromScale);
+
+        if (overflows(returnValue, MAX_SHORT_PRECISION)) {
```

Comment:
was this a bug? If so, please extract to separate commit and add tests

Reply:
I think it was, I have added test for it. Let me separate it out in a different commit.